src/iop/filmicrgb.c

```diff
@@ -1502,6 +1504,363 @@ static inline void filmic_chroma_v2_v3(const float *const restrict in, float *co
+  // when saturation > 1, we go back towards the initial chroma before tone-mapping
+  // when saturation < 0, we amplify the initial -> final chroma change
+  const float delta_chroma = saturation * (chroma_original - chroma_final);
+
+  const int filmic_brightens = (Ych_final[0] > Ych_original[0]);
+  const int filmic_resat = (chroma_original < chroma_final);
+  const int filmic_desat = (chroma_original > chroma_final);
+  const int user_resat = (saturation > 0.f);
+  const int user_desat = (saturation < 0.f);
+
+  chroma_final = (filmic_brightens && filmic_resat)
+                      ? chroma_original // force original lower sat if brightening
+                  : ((user_resat && filmic_desat) || user_desat)
+                      ? chroma_final + delta_chroma // allow resaturation only if filmic desaturated, allow desat anytime
+                      : chroma_final;
+
+  Ych_final[1] = fmaxf(chroma_final / Ych_final[0], 0.f);
+}
+
+// Pipeline and ICC luminance is CIE Y 1931
+// Kirk Ych/Yrg uses CIE Y 2006
+// 1 CIE Y 1931 = 0.945302269001 CIE Y 2006, so we need to adjust that
+#define CIE_Y_1931_to_CIE_Y_2006(x) 0.945302269001f * x
+
+
+static inline void gamut_check_RGB(dt_aligned_pixel_t RGB_in, dt_aligned_pixel_t Ych_in,
+                                   const dt_colormatrix_t matrix_in, const dt_colormatrix_t matrix_out,
+                                   const float display_black, const float display_white)
+{
+  const int max_iter = 8;
+
+  int iter = 0;
+  float max_pix = fmaxf(fmaxf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+  while(max_pix > display_white && iter < max_iter)
+  {
+    // Perform exposure correction to fit back in gamut
+    // This gives us the closest in-gamut pixel at same saturation.
+    dt_aligned_pixel_t RGB_darkened = { 0.f };
+    for_each_channel(c, aligned(RGB_darkened)) RGB_darkened[c] = RGB_in[c] / max_pix;
+    dt_aligned_pixel_t Ych_darkened = { 0.f };
+    pipe_RGB_to_Ych(RGB_darkened, matrix_in, Ych_darkened);
+
+    // Remapping goal:
+    // 1. at constant hue and luminance, march toward the chroma of the closest same-saturation in-gamut color.
+    //    Since same-saturation is darker, its chroma should be much lower than the current.
+    // 2. iterate until convergence.
+    Ych_in[1] = Ych_darkened[1] * Ych_darkened[0] / Ych_in[0];
+    Ych_to_pipe_RGB(Ych_in, matrix_out, RGB_in);
+    max_pix = fmaxf(fmaxf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+    iter++;
+  }
+
+  // Finally, look for negatives.
+  float min_pix = fminf(fminf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+  iter = 0;
+  while(min_pix < display_black && iter < max_iter)
+  {
+    // Perform a black offset to fit back in gamut
+    // This gives us a whitest version of the current color.
+    // It's not saturation-invariant.
+    dt_aligned_pixel_t RGB_brightened = { 0.f };
+    for_each_channel(c, aligned(RGB_brightened))
+      RGB_brightened[c] = RGB_in[c] + fabsf(min_pix);
```

Comment:
Should `display_black` be taken in account here? i.e.

```c
RGB_brightened[c] = RGB_in[c] + MAX(display_black - min_pix, 0.f);
```

Reply:
I scratched my head about that, but black is supposed to be defined as an Y threshold, so we don't have to enforce the black condition on RGB.

Reply:
Doesn't the `while` loop here however enforce exactly that? I.e.

```c
while(min_pix < display_black && iter < max_iter)
{
  ...
}
```

It loops until the smallest pixel value reaches at least `display_black`. I tried to reason about this too when writing a gamut mapping implementation. Isn't it so that the RGB <-> XYZ is normalized such that a RGB tuple `(Y, Y, Y)` has a luminance of Y? The easiest way to enforce a certain boundary luminance is probably trying to get each RGB component within that limit. See also comment above about `display_white`.

Reply:
I don't know. The thing is, if you prevent individual RGB channels to go to 0, you actually prevent not only pure black but also the most saturated colors. Given that the RGB norm is already set between display black and display white by the tone mapping, I think it's overkill to overload that condition here.

Reply:
Yes, I have been wondering about that also, and agree that it's better not to clamp individual channels in order to allow for most saturated colors. But then, the while loop condition should also change:

```c
while(min_pix < 0.f && iter < max_iter)
{
  ...
}
```

Or, in my version of the chroma clipping using closed-form equations, I should change

```c
// Calculate the maximum chroma where RGB components still stay above black point
const float chroma_clipped_to_black = clip_chroma(matrix_out, display_black, Y, cos_h, sin_h);
```

to
```c
const float chroma_clipped_to_black = clip_chroma(matrix_out, 0.f, Y, cos_h, sin_h);
```